File: Python/OneLang/One/Transforms/CollectInheritanceInfo.py
from OneLangStdLib import *
import OneLang.One.Ast.Types as types
import OneLang.One.ITransformer as iTrans

class CollectInheritanceInfo:
    def __init__(self):
        self.name = "CollectInheritanceInfo"
        # C# fix
        self.name = "CollectInheritanceInfo"
    
    def visit_class(self, cls_):
        all_base_iintfs = cls_.get_all_base_interfaces()
        intfs = list(filter(lambda x: x != None, list(map(lambda x: x if isinstance(x, types.Interface) else None, all_base_iintfs))))
        clses = list(filter(lambda x: x != None and x != cls_, list(map(lambda x: x if isinstance(x, types.Class) else None, all_base_iintfs))))
        
        for field in cls_.fields:
            field.interface_declarations = list(filter(lambda x: x != None, list(map(lambda x: next(filter(lambda f: f.name == field.name, x.fields), None), intfs))))
        
        for method in cls_.methods:
            method.interface_declarations = list(filter(lambda x: x != None, list(map(lambda x: next(filter(lambda m: m.name == method.name, x.methods), None), intfs))))
            method.overrides = next(filter(lambda x: x != None, list(map(lambda x: next(filter(lambda m: m.name == method.name, x.methods), None), clses))), None)
            if method.overrides != None:
                method.overrides.overridden_by.append(method)
    
    def visit_files(self, files):
        for file in files:
            for cls_ in file.classes:
                self.visit_class(cls_)

File: PHP/src/OneLang/One/Transforms/CollectInheritanceInfo.php
<?php

namespace OneLang\One\Transforms\CollectInheritanceInfo;

use OneLang\One\Ast\Types\Package;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\Interface_;
use OneLang\One\Ast\Types\SourceFile;
use OneLang\One\ITransformer\ITransformer;

class CollectInheritanceInfo implements ITransformer {
    public $name = "CollectInheritanceInfo";
    
    function __construct() {
        // C# fix
        $this->name = "CollectInheritanceInfo";
    }
    
    function visitClass($cls) {
        $allBaseIIntfs = $cls->getAllBaseInterfaces();
        $intfs = array_values(array_filter(array_map(function ($x) { return $x instanceof Interface_ ? $x : null; }, $allBaseIIntfs), function ($x) { return $x !== null; }));
        $clses = array_values(array_filter(array_map(function ($x) { return $x instanceof Class_ ? $x : null; }, $allBaseIIntfs), function ($x) use ($cls) { return $x !== null && $x !== $cls; }));
        
        foreach ($cls->fields as $field)
            $field->interfaceDeclarations = array_values(array_filter(array_map(function ($x) use ($field) { return \OneLang\Core\ArrayHelper::find($x->fields, function ($f) use ($field) { return $f->name === $field->name; }); }, $intfs), function ($x) { return $x !== null; }));
        
        foreach ($cls->methods as $method) {
            $method->interfaceDeclarations = array_values(array_filter(array_map(function ($x) use ($method) { return \OneLang\Core\ArrayHelper::find($x->methods, function ($m) use ($method) { return $m->name === $method->name; }); }, $intfs), function ($x) { return $x !== null; }));
            $method->overrides = \OneLang\Core\ArrayHelper::find(array_map(function ($x) use ($method) { return \OneLang\Core\ArrayHelper::find($x->methods, function ($m) use ($method) { return $m->name === $method->name; }); }, $clses), function ($x) { return $x !== null; });
            if ($method->overrides !== null)
                $method->overrides->overriddenBy[] = $method;
        }
    }
    
    function visitFiles($files) {
        foreach ($files as $file)
            foreach ($file->classes as $cls)
                $this->visitClass($cls);
    }
}
